### DISCUSSION TITLE
Parallel Operations and Events

### DISCUSSION THREAD

**Discussion by rguillome** [serverlessworkflow/specification#665] [Q&A] [answered]
Hi,

I wonder if there is a way to represent sort of parallele and join stuff, including both operations and events. The use case, would be to orchestrate actions currently runnable by a worker while some other branch should way the incoming of events representing the completion of tasks outside of the workflow engine (like the end of human tasks notified).

**Comment by JBBianchi** [comment]
I guess you could use a parallel [operation state](https://github.com/serverlessworkflow/specification/blob/main/specification.md#operation-state) with mixed functions and events [actions](https://github.com/serverlessworkflow/specification/blob/main/specification.md#Action-Definition) ?

Your question might also be related to https://github.com/serverlessworkflow/specification/issues/677 .

**Reply by rguillome** [reply]
Thanks @JBBianchi : my wish was to parallelize both Operation state and Event state not events actions because it's for example a manual action outside the workflow and my workflow has just to wait for its completion (via a signal/event).

I guess @tsurdilo's answer is the best way today. I will experiment that.

I still wonder why there couldn't have events in Parallele State Branch avoiding the complexity of a Subflow

**Reply by JBBianchi** [reply]
@rguillome I understood your concern, that's one of the concerns/use cases raised in #677. You might also want to have a look at the [Callback state](https://github.com/serverlessworkflow/specification/blob/main/specification.md#callback-state).

**Reply by rguillome** [reply]
Yes thanks @JBBianchi very clear and I agree with you. For the Callback, it could answer some of my use case but not the one that's waiting for an event without calling an external service.

**Comment by tsurdilo** [comment, marked as answer]
If I understand correctly this could be possible right now. You can have branches that run actions and some branches that deal with events. The event handling can be offloaded to a subflow that has event state to deal with the events. Then use the parallel state completionType / numCompleted property for the join part if needed.

**Reply by rguillome** [reply]
I finally manage to test this in a java sample with temporal IO : https://github.com/Zenika/temporalio-serverlessworkflow
After some dirty adaptations (I will make them cleaner in a PR in sample-java), it works. But it's a bit overkill to just add eventState in branches that I think should be available. 
If there is no issues about that (#667  seems more focused on duplicity between event state and consume event in actions), I will add a new one.

**Reply by JBBianchi** [reply]
> I finally manage to test this in a java sample with temporal IO : https://github.com/Zenika/temporalio-serverlessworkflow After some dirty adaptations (I will make them cleaner in a PR in sample-java), it works. But it's a bit overkill to just add eventState in branches that I think should be available. If there is no issues about that (#667 seems more focused on duplicity between event state and consume event in actions), I will add a new one.

I think it's a valid point worth of an issue indeed. Please feel free to open one.
